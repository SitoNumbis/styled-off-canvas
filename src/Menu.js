import React, { useEffect, useContext } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Context from './Context'

const Container = styled.div(({
  background,
  duration,
  isOpen,
  position,
  width
}) => `
  background-color: ${background};
  height: 100%;
  left: ${position === 'left' ? isOpen ? 0 : '-100%' : 'inherit'};
  max-width: ${width};
  overflow-y: auto;
  position: fixed;
  right: ${position === 'right' ? isOpen ? 0 : '-100%' : 'inherit'};
  top: 0;
  transition: ${position} ${duration} ease-in-out;
  width: 100%;
  z-index: 10;
`)

const initalBodyStyle = {
  maxHeight: document.body.style.maxHeight,
  overflow: document.body.style.overflow
}

/**
 * Menu component
 *
 * @param   {string}    className styled-components className for custom styling
 * @returns {component}           React component
 */
const Menu = ({
  children,
  className
}) => {
  const { isOpen, menuBackground, menuDuration, closeOnEsc, onClose, position, width } = useContext(Context)
  // Avoid scrolling on content when the navigation is open
  useEffect(() => {
    const bodyElStyle = document.body.style

    if (isOpen) {
      bodyElStyle.maxHeight = '100%'
      bodyElStyle.overflow = 'hidden'
    } else {
      bodyElStyle.maxHeight = initalBodyStyle.maxHeight
      bodyElStyle.overflow = initalBodyStyle.overflow
    }
  }, [isOpen])

  if (closeOnEsc) {
    useEffect(() => {
      const handleKeyDown = ({ keyCode }) => {
        if (keyCode === 27) onClose()
      }

      isOpen
        ? document.addEventListener('keydown', handleKeyDown)
        : document.removeEventListener('keydown', handleKeyDown)
    }, [isOpen])
  }

  return (
    <Container
      className={className}
      background={menuBackground}
      duration={menuDuration}
      isOpen={isOpen}
      position={position}
      width={width}
    >
      {children}
    </Container>
  )
}

Menu.propTypes = {
  children: PropTypes.element.isRequired,
  className: PropTypes.object
}

export default Menu
