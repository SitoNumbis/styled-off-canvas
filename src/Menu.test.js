import React from 'react'
import renderer from './helper/tests/jest'

import Menu from './Menu'

describe('Menu', () => {
  test('default', () => {
    const tree = renderer(<Menu isOpen><div>content</div></Menu>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('closed', () => {
    const tree = renderer(<Menu isOpen={false}><div>content</div></Menu>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('custom color', () => {
    const tree = renderer(<Menu isOpen color='#000'><div>content</div></Menu>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('custom duration', () => {
    const tree = renderer(<Menu isOpen duration='1s'><div>content</div></Menu>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('position left', () => {
    const tree = renderer(<Menu isOpen position='left'><div>content</div></Menu>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('position left closed', () => {
    const tree = renderer(<Menu isOpen={false} position='left'><div>content</div></Menu>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('onEsc callback', () => {
    const cb = jest.fn()
    const component = renderer(<Menu onEsc={cb} isOpen><div>content</div></Menu>)
    component.update()

    const escKeyDown = new window.KeyboardEvent('keydown', { keyCode: 27 })
    document.dispatchEvent(escKeyDown)

    expect(cb).toHaveBeenCalled()
  })

  test('onEsc callback when closed', () => {
    const cb = jest.fn()
    const component = renderer(<Menu onEsc={cb}><div>content</div></Menu>)
    component.update()

    const escKeyDown = new window.KeyboardEvent('keydown', { keyCode: 27 })
    document.dispatchEvent(escKeyDown)

    expect(cb).not.toHaveBeenCalled()
  })

  test('onEsc callback with another key', () => {
    const cb = jest.fn()
    const component = renderer(<Menu onEsc={cb} isOpen><div>content</div></Menu>)
    component.update()

    const escKeyDown = new window.KeyboardEvent('keydown', { keyCode: 99 })
    document.dispatchEvent(escKeyDown)

    expect(cb).not.toHaveBeenCalled()
  })

  test('body styling', () => {
    const component = renderer(<Menu isOpen><div>content</div></Menu>)

    component.update()
    expect(document.body.style._values).toEqual({
      'max-height': '100%',
      overflow: 'hidden'
    })

    component.update(<Menu><div>content</div></Menu>)
    component.update()
    expect(document.body.style._values).toEqual({})
  })
})
