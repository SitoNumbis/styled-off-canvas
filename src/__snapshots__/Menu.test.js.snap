// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu closed 1`] = `
<div
  className="Menu__Container-mw2r80-0 kszveX"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu custom color 1`] = `
<div
  className="Menu__Container-mw2r80-0 ijZiow"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu custom duration 1`] = `
<div
  className="Menu__Container-mw2r80-0 goDVSy"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu default 1`] = `
<div
  className="Menu__Container-mw2r80-0 goDVSy"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu position left 1`] = `
<div
  className="Menu__Container-mw2r80-0 eHAZXF"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu position left closed 1`] = `
<div
  className="Menu__Container-mw2r80-0 cKcJpo"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;
